In [1]:
import pandas as pd

In [2]:
pd.set_option('display.max_rows',None)          # Mosta tutte le righe

pd.set_option('display.max_columns',None)       # Mostra tutte le colonne

pd.set_option('display.max_info_columns',200)   # df.info() mostra 200 colonne
                                                # (200 è un numero indicativo, che serve per mostrare tutte le colonne
                                                #  se le colonne fossero di più di 200 bisognerebbe inserire un numero maggiore)

In [3]:
df1 = pd.read_csv('../data/raw/crime-2010-2019.csv') # Lettura DataFrame relativo ai crimini 2010-2019
df2 = pd.read_csv('../data/raw/crime-2020-2024.csv') # Lettura DataFrame relativo ai crimini 2020-2024

In [4]:
var1=df1.columns # Creazione della lista delle colonne dei DataFrame 'df1' e 'df2'
var2=df2.columns # per controllare se il nome delle colonne sia lo stesso per entrambi i DataFrame prima del merge

print( f'Le colonne del primo df sono: {var1}')
print( f'Le colonne del secondo df sono: {var2}')

Le colonne del primo df sono: Index(['DR_NO', 'Date Rptd', 'DATE OCC', 'TIME OCC', 'AREA ', 'AREA NAME',
       'Rpt Dist No', 'Part 1-2', 'Crm Cd', 'Crm Cd Desc', 'Mocodes',
       'Vict Age', 'Vict Sex', 'Vict Descent', 'Premis Cd', 'Premis Desc',
       'Weapon Used Cd', 'Weapon Desc', 'Status', 'Status Desc', 'Crm Cd 1',
       'Crm Cd 2', 'Crm Cd 3', 'Crm Cd 4', 'LOCATION', 'Cross Street', 'LAT',
       'LON'],
      dtype='object')
Le colonne del secondo df sono: Index(['DR_NO', 'Date Rptd', 'DATE OCC', 'TIME OCC', 'AREA', 'AREA NAME',
       'Rpt Dist No', 'Part 1-2', 'Crm Cd', 'Crm Cd Desc', 'Mocodes',
       'Vict Age', 'Vict Sex', 'Vict Descent', 'Premis Cd', 'Premis Desc',
       'Weapon Used Cd', 'Weapon Desc', 'Status', 'Status Desc', 'Crm Cd 1',
       'Crm Cd 2', 'Crm Cd 3', 'Crm Cd 4', 'LOCATION', 'Cross Street', 'LAT',
       'LON'],
      dtype='object')


In [5]:
var1==var2  # Confronto del nome delle colonne

array([ True,  True,  True,  True, False,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

In [6]:
df1.columns = df1.columns.str.strip() 
df2.columns = df2.columns.str.strip()

In [7]:
new_col1 = df1.columns
new_col2 = df2.columns

new_col1==new_col2

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True,  True,  True,  True,  True,  True,  True,
        True])

In [8]:
df = pd.concat([df1, df2], ignore_index=True)
print(df.shape)
print(df.info())

(3138031, 28)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3138031 entries, 0 to 3138030
Data columns (total 28 columns):
 #   Column          Dtype  
---  ------          -----  
 0   DR_NO           int64  
 1   Date Rptd       object 
 2   DATE OCC        object 
 3   TIME OCC        int64  
 4   AREA            int64  
 5   AREA NAME       object 
 6   Rpt Dist No     int64  
 7   Part 1-2        int64  
 8   Crm Cd          int64  
 9   Crm Cd Desc     object 
 10  Mocodes         object 
 11  Vict Age        int64  
 12  Vict Sex        object 
 13  Vict Descent    object 
 14  Premis Cd       float64
 15  Premis Desc     object 
 16  Weapon Used Cd  float64
 17  Weapon Desc     object 
 18  Status          object 
 19  Status Desc     object 
 20  Crm Cd 1        float64
 21  Crm Cd 2        float64
 22  Crm Cd 3        float64
 23  Crm Cd 4        float64
 24  LOCATION        object 
 25  Cross Street    object 
 26  LAT             object 
 27  LON             object 
dty

In [10]:
duplicati = df['DR_NO'].duplicated().sum()
print(f"Duplicati su DR_NO: {duplicati}")

Duplicati su DR_NO: 57809


In [11]:
nulli_pct = (df.isnull().sum() / len(df) * 100).round(2)
nulli_pct = nulli_pct[nulli_pct > 0].sort_values(ascending=False)
print("Percentuale di valori nulli per colonna:")
print(nulli_pct)

Percentuale di valori nulli per colonna:
Crm Cd 4          99.99
Crm Cd 3          99.81
Crm Cd 2          93.27
Cross Street      83.71
Weapon Used Cd    66.73
Weapon Desc       66.73
Mocodes           12.15
Vict Sex          10.92
Vict Descent      10.92
Premis Desc        0.02
dtype: float64


In [12]:
df['anno_temp'] = pd.to_datetime(df['DATE OCC'], errors='coerce').dt.year
print("Record per anno:")
print(df['anno_temp'].value_counts().sort_index())
df.drop(columns='anno_temp', inplace=True)

/var/folders/wl/_8d4j_cs57qgchs9ckhqzkyr0000gn/T/ipykernel_1806/4075905120.py:1: UserWarning: Could not infer format, so each element will be parsed individually, falling back to `dateutil`. To ensure parsing is consistent and as-expected, please specify a format.
  df['anno_temp'] = pd.to_datetime(df['DATE OCC'], errors='coerce').dt.year


Record per anno:
anno_temp
2010    209325
2011    200912
2012    201835
2013    192875
2014    195879
2015    168076
2016    283798
2017    231751
2018    229768
2019    218918
2020    199847
2021    209876
2022    235259
2023    232345
2024    127567
Name: count, dtype: int64


## Conclusioni — Fase Plan

**Dimensioni dataset**: 3.138.031 righe × 28 colonne (~670 MB).

**Duplicati**: 57.809 record duplicati su DR_NO (~1.8%). 
Saranno gestiti in fase di cleaning mantenendo l'ultima occorrenza.

**Valori nulli rilevanti**:
- `Crm Cd 2/3/4`: >93% nulli → colonne da eliminare (non rilevanti per le domande guida)
- `Cross Street`: 83% nulli → eliminabile (info ridondante con LOCATION e LAT/LON)
- `Weapon Used Cd / Weapon Desc`: 66% nulli → semantica "nessuna arma", da mantenere
- `Mocodes`: 12% nulli → da valutare (codifica complessa)
- `Vict Sex / Vict Descent`: 11% nulli → ricodificare come "Unknown"

**Anomalie temporali identificate**:
- Anni 2015–2016 mostrano valori anomali (calo nel 2015, picco nel 2016), 
  probabilmente legati alla transizione del sistema di classificazione LAPD.
- Anno 2024 incompleto (127.567 record vs ~230k attesi): il dataset risulta 
  congelato a metà anno. Da escludere dalle analisi di trend annuali.

**Tipi di dato da convertire**:
- `Date Rptd`, `DATE OCC` → datetime
- `TIME OCC` → orario formattato
- `LAT`, `LON` → float (attualmente object, da indagare)
- `Vict Age` → gestire valori sentinella (0, negativi)

**Decisioni operative per la fase Analyze**:
- Periodo di riferimento: 2015–2024 (come da PACE doc)
- Anno 2024 escluso dalle visualizzazioni di trend annuali per incompletezza
- Anomalie 2015–2016 documentate e segnalate nei grafici